File: iOS-Lab1/CalBrain.m
//
//  CalBrain.m
//  Sep21_Calculator
//
//  Created by mays on 2023-09-21.
//

#import "CalBrain.h" // added the class header file



@interface CalBrain()

//private property of items : nsmutableArray
@property(nonatomic,strong)NSMutableArray *items;
@end
@implementation CalBrain
// lazy load
-(NSMutableArray*)items{
    // if intem index not empty
    if (_items == nil){

        _items =  [[NSMutableArray alloc]init];
    }
        return _items;
}

-(void)pushItem:(double)item{
    //add this number to my items list
   
    [self.items addObject:@(item)];
   
}
// calculate function
-(double)calculate:(NSString *)opr{
    
  
  
    //
    //check what opr is + or - * /
    // if oporatoer is + symbol then add two numbers
    if([opr isEqualToString:@"+"]){
         
        return [self popItem] +[self popItem];
        // if oporatoer is -  symbol then subtract two numbers
    }else if ([opr isEqualToString:@"-"]){
        return [self popItem] -[self popItem];
        // if oporatoer is *  symbol then multiply two numbers
    }else if ([opr isEqualToString:@"*"]){
        return [self popItem] * [self popItem];
        // if oporatoer is / symbol then divide two numbers
    }else if ([opr isEqualToString:@"/"]){
        // Store the last value poped
        double num2 = [self popItem];
        // Store the first value poped
        double num1 = [self popItem];
        if(num2 ==0){
            // print not allowed to divide
            NSLog(@"can not divied by zero");
            // return NAN
            return  NAN;
        }
        // return the division value
        return num1 / num2;
    }else{
        // print in case invalid oprator then return NAN
        NSLog(@"Error: Invalid operator.");
        
    }
    return NAN;
}


-(double)popItem{
    //remove and return the last item of the array
   // Check
    if([self.items count] > 0){
        // Store the last last item from the array in var lastItem
        double lastItem =[[self.items lastObject] doubleValue];
        // Remove the last item for the array
        [self.items removeLastObject];
        return  lastItem;
    }else{
        return  NAN;
    }
    
}



@end

File: iOS/Swift/Lab1/Calculator/Calculator_MahaB/CalBrain.swift
//
//  CalBrain.swift
//  Calculator_MahaB
//
//  Created by Maha Basheikh on 2023-10-05.
//

import UIKit

class CalBrain: NSObject {
    // lazy loading the array
    private lazy var items = NSMutableArray()
    
    // method to push numbers to the array
    func pushItem(number: Double) {
        //self.items.addObjects(from: number)
        
        // add a number to the array
        items.add(number)
    }
    
    // method to pop or take out a number from the array to use
    func popItem() -> Double {
        if items.count > 0 { // if the array is not empty
            //var lastItem: Double
            
            //unwrap the optional as a double
            if let lastItem = items.lastObject as? Double {
                items.removeLastObject() // take out the last object from the array
                
                return lastItem //return it to the caller
            }
        }
        else {
            return -1.0 // if the array is empty, return a -1, signaling an error
        }
              
        return -1.0 // if the array is empty, return a -1, signalling an error
    }
    
    // method to calculate the operation
    func calculate(opr: String) -> Double? {
        if opr == "+" {
            return popItem() + popItem() // addition
        }
        else if opr == "-" {
            return popItem() - popItem() // subtraction
        }
        else if opr == "✖️" {
            return popItem() * popItem() // multiplication
        }
        else if opr == "÷" { // division
            
            // declaring the numerator and denominator
            var numerator: Double
            var denominator: Double
            
            // popping the numberator and denominator
            denominator = popItem()
            numerator = popItem()
            
            if denominator == 0 { // We are making sure to let the user know they can't divide from zero and exiting
                print("Cannot divide by zero!")
                return nil
            }
            return numerator / denominator // returning the result of the calculation
        }
        else {
            print("Invalid operator") // if an invalid is somehow entered
        }
        return -1.0
    }
    
}

